icnsutils/icnsutil.1
.TH ICNSUTIL "1" "Oct 2012" "icnsutil 0.1" "User Commands"
.SH NAME
icnsutil \- Utility to convert between '.iconset' and '.icns' files.
.SH SYNOPSIS
.B iconutil -c {icns | iconset} [-o \fIfile\fR] \fIfile\fR
.SH DESCRIPTION
icnsutil is intended to be a functional clone of the Darwin iconutil,
 which converts between '.iconset' and '.icns' files. The tool takes a
 single source '.icns' file or '.iconset' and converts it to either a
 '.icns' or '.iconset' depending on the value of the -c flag's
 argument.
.P
 It is possible to specify the name of the output file by passing the
 file path as the argument to the -o flag. If -o is not set iconutil
 will write converted '.icns' file or '.iconset' to the same directory
 as the source file using the same file name with the correct extension
 for the output type.
.SH AUTHOR
Written by Mathew Eis
.SH COPYRIGHT
Copyright \(co 2012 Mathew Eis
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.br


NEWS
Release 0.8.0  (Current)
# Sourceforge SVN rev 170 - current
- libicns version 2.8.2 (per libtool)
- fixed potentially dangerous parsing error
- improved icns2png's understanding of 'icnV' type
- added awareness of 'TOC ' type to libicns/icns2png
- added support for PNG icns elements
- added support for icon variant type
- fix for alignment/bus errors on ARM/Sparc 

Release 0.7.1  (09/03/2009)
# Sourceforge SVN rev 165 - 170
- libicns version 2.7.1 (per libtool)
- 

Release 0.7.0  (07/28/2009)
# Sourceforge SVN rev 163 - 165
- libicns version 2.7.1 (per libtool)
- Fixed licensing concerns

Release 0.6.2  (05/10/2009)
# Sourceforge SVN rev 154 - 162
- libicns version 2.6.1 (per libtool)
- Changed icns_type_t from struct to uint32_t
- Updated all code relevant to the icns_type_t change
- Added a function for creating NULL terminated type strings

Release 0.6.1  (02/05/2009)
# Sourceforge SVN rev 152 - 153
- libicns version 2.6.1 (per libtool)
- Removed debug code from png2icns
- Added CDEF block to jp2 writing code
- Modified icns_set_element_in_family to sort newly inserted elements
- Added icns_type_to_uint32 to libicns internals (for faster type checking)
- Added icns_get_element_order to libicns internals (for sorting elements)
- Minor performance improvements

Release 0.6.0  (01/05/2009)
# Sourceforge SVN rev 149 - 151
- libicns version 2.5.1 (per libtool)
- Fixed bug in with global variable - SF Bug 2139996
- Fixed bug in png2icns; info was not properly updated after transforms

Release 0.5.9  (10/09/2008)
# Sourceforge SVN rev 138 - 148
# libicns version 2.4.1 (per libtool)
- Fixed several small bugs
- Fixed major bug in RLE compressor
- Changed icns2png to save extracted files to current directory by default
- Completed write support for 256x256 & 512x512 via Jasper

Release 0.5.8  (06/12/2008)
# Sourceforge SVN rev 135 - 137
# libicns version 2.3.1 (per libtool) - (NOTE: fix to 2.3.1 at SVN rev 137)
- Fixed licensing issues with FSF address
- Fixed incorrect libicns/libtool versioning

Release 0.5.7  (06/10/2008)
# Sourceforge SVN rev 118 - 134
# libicns version 2.0.1 (per libtool) - (NOTE: should have changed to 2.2.1)
- Updated to use autoconf 2.61, automake 1.10
- Added debain packaging to SVN tree (NOTE: Maintainer use only!)
- Moved libicns/ to src/ for more conventional structure
- Added png2icns utility
- Added confgure check for libjasper and/or libopenjpeg
- Added manpages for png2icns and icontainer2icns
- Prevented libicns from exporting internal-only symbols
- Fixed some compiler errors regarding pointers

Release 0.5.6  (04/10/2008)
# Sourceforge SVN rev 109 - 117
# libicns version 2.0.1 (per libtool) - (NOTE: should have changed to 2.1.1)
- Several write support bugfixes (write support was broken up to this point)
- Minor code clarifications
- Updated icns2png manpage


Release 0.5.5  (03/17/2008)
# Sourceforge SVN rev 102 - 108
# libicns version 2.0.1 (per libtool)
- Converted project to use GNU automake and friends
- Converted all files to UTF-8
- Added multiple batch extraction to icns2png
- Added png output path to icns2png
- Enhanced icns2png for Mac OS X (accesses resource fork now)
- Added HTML API documentation

Release 0.5.2  (03/11/2008)
# Sourceforge SVN rev 93 - 101
- Restructured code tree to seperate icns2png from libicns
- Minor bugfixes

Release 0.5.1  (03/07/2008)
# Sourceforge SVN rev 88 - 92
- Added Jasper JP2 decoder
- Minor bugfixes

Release 0.5.0  (03/05/2008)
# Sourceforge SVN rev 75 - 87
- Added basic API documentation
- Dramatially enhanced handling of endian issues
- Added some minor debug checks
- Changed data types to use stdint.h

Release 0.4.1  (02/28/2008)
# Sourceforge SVN rev 72 - 74
- Fixed libtool issue under Darwin
- Several minor bugfixes

Release 0.3.0  (02/26/2008)
# Sourceforge SVN rev 63 - 71
- Fixed numerous bugs
- Split icns.c into manageable portions
- Added function to retrieve image with mask from icon family
- Modified endian swap routines for easier debugging
- Added routines for accessing unaligned memory
- Made minor code changes for consistency within libicns

Release 0.1.1  (02/20/2008)
# Sourceforge SVN rev 62
- Fixed dynamic library build issues under OS X
- Fixed bugs when uncompressing RLE data in sizes < 128
- Other minor bugfixes
- Added changelog
- Added versioning specifications

Release 0.1.0  (02/18/2008)
- first major release of libicns with write support
- added read support for 1,4,8 bit icons
- removed all possible Apple copyrighted stuff
- added support to modify the icns structure
- added support for 32-bit image manipulation
- added support to do internal rle compression
- added file read/write support using the library directly
- fixed Mac OS X build

Release 0.0.9  (02/14/2008)
- merged in test icns files from original source tree
- added TODO for libicns developers
- finished breakdown of libicns and icns2png
- merged all headers required by libicns into icns.h
- moved jpeg decoding routines into icns.c
- established libicns development naming conventions
- added DEVNOTES

Release 0.0.8  (02/13/2008)
- merged forked source trees from chenxio and baghira
- preliminary split of source into libicns and icns2png

Release 0.0.7  (01/16/2008)
- forked by chenxiao.zhao at Google Code
- added 256x256 and 512x512 support via openjpeg
- added command-line parameters

Release 0.0.6  (08/30/2007)
- forked by baghira at CLI-Apps
- added "NG" icns support
- added icontainter2icns utility

Release 0.0.5 (01/28/2003)
- Fixed bug in resource fork parsing

Release 0.0.4 (01/23/2003)
- Fixed a bug with filenames that was causing segfaults

Release 0.0.3 (12/18/2002)
- Much improved parsing of the icons within resource forks and resource files.
- Much improved MacBinary support.
- Updated README with unstuff information

Release 0.0.2 (01/11/2002)
- Inital public release



icnsutils/icnsutil.c
/*
 * icnsutil
 *
 * Copyright (C) 2008 Julien BLACHE <jb@jblache.org>
 * Copyright (C) 2012 Mathew Eis <mathew@eisbox.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <string.h>
#include <stdint.h>

#include <errno.h>

#include <png.h>
#include <icns.h>

#define	FALSE	0
#define	TRUE	1

#if PNG_LIBPNG_VER >= 10209
 #define PNG2ICNS_EXPAND_GRAY 1
#endif

static const char *icon_names[] = {
	"icon_16x16.png",
	"icon_16x16@2x.png",
	"icon_32x32.png",
	"icon_32x32@2x.png",
	"icon_128x128.png",
	"icon_128x128@2x.png",
	"icon_256x256.png",
	"icon_256x256@2x.png",
	"icon_512x512.png",
	"icon_512x512@2x.png",
	NULL
};

static int read_png(FILE *fp, png_bytepp buffer, int32_t *bpp, int32_t *width, int32_t *height)
{
	png_structp png_ptr;
	png_infop info;
	png_uint_32 w;
	png_uint_32 h;
	png_bytep *rows;

	int bit_depth;
	int32_t color_type;

	int row;
	int rowsize;

	png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
	if (png_ptr == NULL)
		return FALSE;

	info = png_create_info_struct(png_ptr);
	if (info == NULL)
	{
		png_destroy_read_struct(&png_ptr, NULL, NULL);
		return FALSE;
	}

	if (setjmp(png_jmpbuf(png_ptr)))
	{
		png_destroy_read_struct(&png_ptr, &info, NULL);
		return FALSE;
	}

	png_init_io(png_ptr, fp);

	png_read_info(png_ptr, info);
	png_get_IHDR(png_ptr, info, &w, &h, &bit_depth, &color_type, NULL, NULL, NULL);

	switch (color_type)
	{
		case PNG_COLOR_TYPE_GRAY:
			#ifdef PNG2ICNS_EXPAND_GRAY
			png_set_expand_gray_1_2_4_to_8(png_ptr);
			#else
			png_set_gray_1_2_4_to_8(png_ptr);
			#endif

			if (bit_depth == 16) {
				png_set_strip_16(png_ptr);
				bit_depth = 8;
			}

			png_set_gray_to_rgb(png_ptr);
			png_set_add_alpha(png_ptr, 0xff, PNG_FILLER_AFTER);
			break;

		case PNG_COLOR_TYPE_GRAY_ALPHA:
			#ifdef PNG2ICNS_EXPAND_GRAY
			png_set_expand_gray_1_2_4_to_8(png_ptr);
			#else
			png_set_gray_1_2_4_to_8(png_ptr);
			#endif

			if (bit_depth == 16) {
				png_set_strip_16(png_ptr);
				bit_depth = 8;
			}

			png_set_gray_to_rgb(png_ptr);
			break;

		case PNG_COLOR_TYPE_PALETTE:
			png_set_palette_to_rgb(png_ptr);

			if (png_get_valid(png_ptr, info, PNG_INFO_tRNS))
				png_set_tRNS_to_alpha(png_ptr);
			else
				png_set_add_alpha(png_ptr, 0xff, PNG_FILLER_AFTER);
			break;

		case PNG_COLOR_TYPE_RGB:
			if (bit_depth == 16) {
				png_set_strip_16(png_ptr);
				bit_depth = 8;
			}

			png_set_add_alpha(png_ptr, 0xff, PNG_FILLER_AFTER);
			break;

		case PNG_COLOR_TYPE_RGB_ALPHA:
			if (bit_depth == 16) {
				png_set_strip_16(png_ptr);
				bit_depth = 8;
			}

			break;
	}

	*width = w;
	*height = h;
	*bpp = bit_depth * 4;

	png_read_update_info(png_ptr, info);

	rowsize = png_get_rowbytes(png_ptr, info);
	rows = malloc (sizeof(png_bytep) * h);
	*buffer = malloc(rowsize * h + 8);

	rows[0] = *buffer;
	for (row = 1; row < h; row++)
	{
		rows[row] = rows[row-1] + rowsize;
	}

	png_read_image(png_ptr, rows);
	png_destroy_read_struct(&png_ptr, &info, NULL);

	free(rows);

	return TRUE;
}

static int add_png_to_family(icns_family_t **iconFamily, char *pngname)
{
	FILE *pngfile;

	int icnsErr = ICNS_STATUS_OK;
	icns_image_t icnsImage;
	icns_image_t icnsMask;
	icns_type_t iconType;
	icns_type_t maskType;
	icns_icon_info_t iconInfo;

	icns_element_t *iconElement = NULL;
	icns_element_t *maskElement = NULL;
	char iconStr[5] = {0,0,0,0,0};
	char maskStr[5] = {0,0,0,0,0};
	int iconDataOffset = 0;
	int maskDataOffset = 0;

	png_bytep buffer;
	int width, height, bpp;

	pngfile = fopen(pngname, "rb");
	if (pngfile == NULL)
	{
		fprintf(stderr, "Could not open '%s' for reading: %s\n", pngname, strerror(errno));
		return FALSE;
	}

	if (!read_png(pngfile, &buffer, &bpp, &width, &height))
	{
		fprintf(stderr, "Failed to read PNG file\n");
		fclose(pngfile);

		return FALSE;
	}

	fclose(pngfile);

	icnsImage.imageWidth = width;
	icnsImage.imageHeight = height;
	icnsImage.imageChannels = 4;
	icnsImage.imagePixelDepth = 8;
	icnsImage.imageDataSize = width * height * 4;
	icnsImage.imageData = buffer;

	iconInfo.isImage = 1;
	iconInfo.iconWidth = icnsImage.imageWidth;
	iconInfo.iconHeight = icnsImage.imageHeight;
	iconInfo.iconBitDepth = bpp;
	iconInfo.iconChannels = (bpp == 32 ? 4 : 1);
	iconInfo.iconPixelDepth = bpp / iconInfo.iconChannels;

	iconType = icns_get_type_from_image_info(iconInfo);
	maskType = icns_get_mask_type_for_icon_type(iconType);

	icns_type_str(iconType,iconStr);
	icns_type_str(maskType,maskStr);

	/* Only convert the icons that match sizes icns supports */
	if (iconType == ICNS_NULL_TYPE)
	{
		fprintf(stderr, "Bad dimensions: PNG file '%s' is %dx%d\n", pngname, width, height);
		free(buffer);

		return FALSE;
	}

	if (bpp != 32)
	{
		fprintf(stderr, "Bit depth %d unsupported in '%s'\n", bpp, pngname);
		free(buffer);

		return FALSE;
	}

	icns_set_print_errors(0);
	if (icns_get_element_from_family(*iconFamily, iconType, &iconElement) == ICNS_STATUS_OK)
	{
		icns_set_print_errors(1);

		fprintf(stderr, "Duplicate icon element of type '%s' detected (%s)\n", iconStr, pngname);
		free(buffer);

		return FALSE;
	}
	
	icns_set_print_errors(1);
	
	if( (iconType != ICNS_1024x1024_32BIT_ARGB_DATA) && (iconType != ICNS_512x512_32BIT_ARGB_DATA) && (iconType != ICNS_256x256_32BIT_ARGB_DATA) )
	{
		printf("Using icns type '%s', mask '%s' for '%s'\n", iconStr, maskStr, pngname);
	}
	else
	{
		printf("Using icns type '%s' (ARGB) for '%s'\n", iconStr, pngname);
	}
	
	icnsErr = icns_new_element_from_image(&icnsImage, iconType, &iconElement);
	
	if (iconElement != NULL)
	{
		if (icnsErr == ICNS_STATUS_OK)
		{
			icns_set_element_in_family(iconFamily, iconElement);
		}
		free(iconElement);
	}

	if( (iconType != ICNS_1024x1024_32BIT_ARGB_DATA) && (iconType != ICNS_512x512_32BIT_ARGB_DATA) && (iconType != ICNS_256x256_32BIT_ARGB_DATA) )
	{
		icns_init_image_for_type(maskType, &icnsMask);

		iconDataOffset = 0;
		maskDataOffset = 0;
	
		while ((iconDataOffset < icnsImage.imageDataSize) && (maskDataOffset < icnsMask.imageDataSize))
		{
			icnsMask.imageData[maskDataOffset] = icnsImage.imageData[iconDataOffset+3];
			iconDataOffset += 4; /* move to the next alpha byte */
			maskDataOffset += 1; /* move to the next byte */
		}

		icnsErr = icns_new_element_from_mask(&icnsMask, maskType, &maskElement);

		if (maskElement != NULL)
		{
			if (icnsErr == ICNS_STATUS_OK)
			{
				icns_set_element_in_family(iconFamily, maskElement);
			}
			free(maskElement);
		}
		
		icns_free_image(&icnsMask);
	}

	free(buffer);

	return TRUE;
}

int usage(void)
{
		printf("Usage: icnsutil -c {icns | iconset} [-o file] file\n");
		exit(1);
		return 1;
}

int iconset_to_icns(char *srcfile, char *dstfile)
{
	FILE *icnsfile;
	icns_family_t	*iconFamily;
	char *pngfile = NULL;
	char *outfile = NULL;
	int	srclen = strlen(srcfile);
	int i = 0;
	
	pngfile = malloc(srclen + 32);
	strncpy(&pngfile[0],&srcfile[0],srclen);
	pngfile[srclen] = '/';
	
	if(dstfile == NULL) {
		int srcstart = srclen - 1;
		
		while(srcfile[srcstart] != '/' && srcfile[srcstart] != '.' && srcstart > 0)
			srcstart--;
			
		if(srcfile[srcstart] != '.' || srcstart == 0)
			srcstart = srclen;
			
		outfile = malloc(srclen + 6);
		strncpy(&outfile[0],&srcfile[0],srcstart);
		strncpy(&outfile[srcstart],".icns",6);
		dstfile = outfile;
	}
	
	icnsfile = fopen (dstfile, "wb+");
	if (icnsfile == NULL)
	{
		fprintf (stderr, "Could not open '%s' for writing: %s\n", dstfile, strerror(errno));
		goto cleanup;
	}
	
	icns_create_family(&iconFamily);
	
	while(icon_names[i] != NULL) {
		strcpy(&pngfile[srclen + 1],icon_names[i]);
		printf("adding png: %s\n",pngfile);
		add_png_to_family(&iconFamily,pngfile);
		i++;
	}

	if (icns_write_family_to_file(icnsfile, iconFamily) != ICNS_STATUS_OK)
	{
		fprintf(stderr, "Failed to write icns file\n");
		fclose(icnsfile);
		goto cleanup;
	}

	fclose(icnsfile);

	printf("Saved icns file to %s\n",dstfile);
	
cleanup:

	if(iconFamily != NULL)
		free(iconFamily);
		
	if(outfile != NULL)
		free(outfile);
		
	if(pngfile != NULL)
		free(pngfile);
		
	return 0;
}

int icns_to_iconset(char *srcfile, char *dstfile)
{
	char *outfile = NULL;
	int	srclen = strlen(srcfile);
	
	if(dstfile == NULL) {
		int srcstart = srclen - 1;
		
		while(srcfile[srcstart] != '/' && srcfile[srcstart] != '.' && srcstart > 0)
			srcstart--;
			
		if(srcfile[srcstart] != '.' || srcstart == 0)
			srcstart = srclen;
			
		outfile = malloc(srclen + 9);
		strncpy(&outfile[0],&srcfile[0],srcstart);
		strncpy(&outfile[srcstart],".iconset",9);
		dstfile = outfile;
	}
	
	printf("src: %s\n",srcfile);
	printf("dst: %s\n",dstfile);

	return 0;
}

int main(int argc, char **argv)
{
	int (*conv_fn)(char *,char *);
	
	if (argc < 4)
		usage();
		
	if ( strcmp(argv[1],"-c") != 0)
		usage();

	if ( strcmp(argv[2],"icns") == 0)
		conv_fn = &iconset_to_icns;
	else if ( strcmp(argv[2],"iconset") == 0)
		conv_fn = &icns_to_iconset;
	else
		usage();
	
	if (strcmp(argv[3],"-o") == 0) {
		return (*conv_fn)(argv[5],argv[4]);
	} else {
		return (*conv_fn)(argv[3],NULL);
	}
	
	return 0;
}
